unused/implements.go

```diff
@@ -72,11 +75,85 @@ func implements(V types.Type, T *types.Interface, msV *types.MethodSet) ([]*type
 			return nil, false
 		}
 
-		if !types.Identical(f.Type(), m.Type()) {
+		if !c.methodIsCompatible(f, m) {
 			return nil, false
 		}
 
 		sels = append(sels, sel)
 	}
 	return sels, true
 }
+
+type methodsChecker struct {
+	typeParams map[*types.TypeParam]types.Type
+}
+
+func newMethodChecker() methodsChecker {
+	return methodsChecker{
+		typeParams: nil, // want to allocate lazily
+	}
+}
+
+// Currently, this doesn't support methods like `foo(x []T)`.
+func (c *methodsChecker) methodIsCompatible(implFunc *types.Func, interfaceFunc *types.Func) bool {
+	if types.Identical(implFunc.Type(), interfaceFunc.Type()) {
+		return true
+	}
+	implSig, implOk := implFunc.Type().(*types.Signature)
+	interfaceSig, interfaceOk := interfaceFunc.Type().(*types.Signature)
+	if !implOk || !interfaceOk {
+		// probably not reachable. handle conservatively.
+		return false
+	}
+
+	if !c.typesAreCompatible(implSig.Params(), interfaceSig.Params()) {
+		return false
+	}
+
+	if !c.typesAreCompatible(implSig.Results(), interfaceSig.Results()) {
+		return false
+	}
+
+	return true
+}
+
+func (c *methodsChecker) typesAreCompatible(implTypes, interfaceTypes *types.Tuple) bool {
+	if implTypes.Len() != interfaceTypes.Len() {
+		return false
+	}
+	for i := 0; i < implTypes.Len(); i++ {
+		if !c.typeIsCompatible(implTypes.At(i).Type(), interfaceTypes.At(i).Type()) {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *methodsChecker) typeIsCompatible(implType, interfaceType types.Type) bool {
+	if types.Identical(implType, interfaceType) {
+		return true
+	}
+	tp, ok := interfaceType.(*types.TypeParam)
+	if !ok {
+		return false
+	}
+	if c.typeParams == nil {
+		c.typeParams = make(map[*types.TypeParam]types.Type)
+	}
+	if c.typeParams[tp] == nil {
+		if !satisfiesConstraint(implType, tp) {
+			return false
+		}
+		c.typeParams[tp] = implType
+		return true
+	}
+	return types.Identical(c.typeParams[tp], implType)
+}
+
+func satisfiesConstraint(t types.Type, tp *types.TypeParam) bool {
+	bound, ok := tp.Constraint().Underlying().(*types.Interface)
+	if !ok {
+		panic("unexpected failure on type assertion (types.Type to *types.Interface)")
```

Comment:
Just do `bound := tp.Constraint().Underlying().(*types.Interface)` — we don't need to check for ok just to manually panic — the panic we get on the failed type assertion is fine.